### PR TITLE
Automatically evaluate checkpoints after copying to NFS

### PR DESCRIPTION
**Pull request by Xirider**
This PR adds the ability to automatically evaluate checkpoints after copying to NFS.

- After the asynchronous copy of a checkpoint to NFS, we check periodically whether all checkpoint parts have finished copying.
- Once all parts are done, we run the eval script defined in `nfs_eval_script_path` for this checkpoint

Add these lines to the internal sweep script to define the eval script path:
https://www.internalfb.com/phabricator/paste/view/P576097431

Testing:
I tested this with a dummy eval script, as the Punit's real eval script should only be used for large runs. I would recommend to check first whether the eval script alone runs correctly on our frozen branch.

### CODE REVIEW

File: metaseq/cli/train.py
Comment (Xirider) [inline metaseq/cli/train.py:503]: Noob question: When running on many nodes, does this only run once?

File: metaseq/cli/train.py
Comment (ruanslv) [inline metaseq/cli/train.py:500]: Could we move this logic to a separate method and delete the existing `_run_evaluations` if it's not used anymore?

File: metaseq/cli/train.py
Comment (ruanslv) [reply metaseq/cli/train.py:503]: Yes

File: metaseq/dataclass/configs.py
Comment (ruanslv) [inline metaseq/dataclass/configs.py:563]: We are calling these flags `cloud` but we are not using them for blob nor S3. Could we either make the logic generic or call them `nfs_` instead?

Comment (ruanslv) [reply metaseq/dataclass/configs.py:563]: Also, if we are adding new flags can we remove the `evaluate_` ones added previously? https://github.com/facebookresearch/metaseq/pull/411

File: metaseq/cli/train.py
Comment (ruanslv) [inline metaseq/cli/train.py:504]: This logic is brittle, could we rely on `num_updates` instead? Also we probably want evals at the end of the training run? See approach in https://github.com/facebookresearch/metaseq/pull/411/files

File: metaseq/cli/train.py
Comment (ruanslv) [inline metaseq/cli/train.py:525]: I don't feel particularly strong either way, but @stephenroller had suggested that we use a python function instead, see comments in https://github.com/facebookresearch/metaseq/pull/411/files. Might be easier for version control?

**Comment by ruanslv** [comment]
Thanks for taking this on! I've added a bunch of comments related to previous PR: https://github.com/facebookresearch/metaseq/pull/411/files

I feel perhaps reusing `_run_evaluations` method and just updating the wait criteria (`sleep` vs `monitored_barrier`) may be easier, but your call if you want to replace it instead. We should keep just one evals method around.

**Review by ruanslv — CHANGES_REQUESTED** [review]

File: metaseq/cli/train.py
Comment (stephenroller) [inline metaseq/cli/train.py:545]: warning? error?

File: metaseq/cli/train.py
Comment (stephenroller) [reply metaseq/cli/train.py:525]: yes plz

File: metaseq/cli/train.py
Comment (stephenroller) [inline metaseq/cli/train.py:531]: Don't we submit a job? and do this async?

File: metaseq/dataclass/configs.py
Comment (Xirider) [reply metaseq/dataclass/configs.py:563]: You are right. It is is also coupled to NFS, as we are checking for existence of files, and the eval script won't work on non NFS files. 
- will change it to `nfs_`.
- remove evaluate flag

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py:525]: Do you mean that instead of running a bash script, we run a python function? Mhm, but we are currently using Punit's bash script for these evals. So we might need to translate that into python first  and validate it again. Not sure that is worth it?

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py:531]: What do mean by this?
As we are inside a new Thread for the copy to NFS, all this is done async from the training process already. The bash script is started and submits some jobs, then waits for checkpoint consolidation, submits more jobs for evals.

File: metaseq/cli/train.py
Comment (ruanslv) [reply metaseq/cli/train.py:525]: Yea if we are already using the bash script everywhere then I agree to keep it this way, at least for now

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py]: - updated to use num_updates
- kept the logic for determining when to evaluate outside, as unlike your implementation its only for nfs
- I'm not sure we actually want to have an extra flag for evals at the end of training. Susan previously wanted me to remove similar logic

File: metaseq/cli/train.py
Comment (ruanslv) [inline metaseq/cli/train.py:531]: Did you mean to use `should_stop` here as an additional check? 

I find it weird not having a flag to control whether we run at the end, are you just going to turn it on by default then? It seems clear to me that we would want to evaluate last checkpoint

cc @suchenzang Peter mentioned you argued in favour of not having flags for last_checkpoint?

**Review by ruanslv — APPROVED** [review]
Overall looks good, added some comments for small adjustments but approving to unblock

File: metaseq/cli/train.py
Comment (suchenzang) [reply metaseq/cli/train.py:531]: yeah hopefully no flag and default to just checkpoint the last checkpoint (which we should be doing now since we checkpoint every epoch)? is there a case where the last step is not also the end of an epoch/data shard?

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py:533]: Good catch! Yes, will change that.

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py:541]: Yeah, I will explain this also in a comment. In the NFS copy code above we are first saving the checkpoint part with a leading "_", and once the copying is finished, we rename it by removing the "_". This way we make sure we don't consider half saved checkpoints.

File: metaseq/cli/train.py
Comment (Xirider) [reply metaseq/cli/train.py]: - Added option 1: eval last checkpoint by default, assuming nfs_eval_script_path is set.
- renamed as suggested to training_finished, that makes it clearer!

**Pull request merged**
